// preact-cli/lib/lib/webpack/webpack-client-config.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _path = require('path');

var _minimatch = require('minimatch');

var _webpack = require('@webpack-blocks/webpack2');

var _devServer = require('@webpack-blocks/dev-server2');

var _devServer2 = _interopRequireDefault(_devServer);

var _htmlWebpackPlugin = require('html-webpack-plugin');

var _htmlWebpackPlugin2 = _interopRequireDefault(_htmlWebpackPlugin);

var _htmlWebpackExcludeAssetsPlugin = require('html-webpack-exclude-assets-plugin');

var _htmlWebpackExcludeAssetsPlugin2 = _interopRequireDefault(_htmlWebpackExcludeAssetsPlugin);

var _scriptExtHtmlWebpackPlugin = require('script-ext-html-webpack-plugin');

var _scriptExtHtmlWebpackPlugin2 = _interopRequireDefault(_scriptExtHtmlWebpackPlugin);

var _copyWebpackPlugin = require('copy-webpack-plugin');

var _copyWebpackPlugin2 = _interopRequireDefault(_copyWebpackPlugin);

var _swPrecacheWebpackPlugin = require('sw-precache-webpack-plugin');

var _swPrecacheWebpackPlugin2 = _interopRequireDefault(_swPrecacheWebpackPlugin);

var _pushManifest = require('./push-manifest');

var _pushManifest2 = _interopRequireDefault(_pushManifest);

var _webpackBaseConfig = require('./webpack-base-config');

var _webpackBaseConfig2 = _interopRequireDefault(_webpackBaseConfig);

var _prerender = require('./prerender');

var _prerender2 = _interopRequireDefault(_prerender);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = env => {
	var _helpers = (0, _webpackBaseConfig.helpers)(env);

	let isProd = _helpers.isProd,
	    src = _helpers.src;


	return _webpack.createConfig.vanilla([(0, _webpackBaseConfig2.default)(env), (0, _webpack.entryPoint)({
		'bundle': (0, _path.resolve)(__dirname, './../entry'),
		'polyfills': (0, _path.resolve)(__dirname, './polyfills')
	}), (0, _webpack.setOutput)({
		path: env.dest,
		publicPath: '/',
		filename: isProd ? "[name].[chunkhash:5].js" : "[name].js",
		chunkFilename: '[name].chunk.[chunkhash:5].js'
	}), (0, _webpack.customConfig)({
		module: {
			loaders: [{
				test: /\.jsx?$/,
				include: [(0, _minimatch.filter)(src('routes') + '/{*.js,*/index.js}'), (0, _minimatch.filter)(src('components') + '/{routes,async}/{*.js,*/index.js}')],
				loader: (0, _path.resolve)(__dirname, './async-component-loader'),
				options: {
					name(filename) {
						let relative = filename.replace(src('.'), '');
						let isRoute = filename.indexOf('/routes/') >= 0;

						return isRoute ? 'route-' + relative.replace(/(^\/(routes|components\/(routes|async))\/|(\/index)?\.js$)/g, '') : false;
					},
					formatName(filename) {
						let relative = filename.replace(src('.'), '');

						return relative.replace(/(^\/(routes|components\/(routes|async))\/|(\/index)?\.js$)/g, '');
					}
				}
			}]
		}
	}), isProd && (0, _webpack.performance)(Object.assign({
		maxAssetSize: 200 * 1000,
		maxEntrypointSize: 200 * 1000,
		hints: 'warning'
	}, env.pkg.performance || {})), (0, _webpack.addPlugins)([new _copyWebpackPlugin2.default([...((0, _webpackBaseConfig.exists)(src('manifest.json')) ? [{ from: 'manifest.json' }] : [{
		from: (0, _path.resolve)(__dirname, '../../resources/manifest.json'),
		to: 'manifest.json'
	}, {
		from: (0, _path.resolve)(__dirname, '../../resources/icon.png'),
		to: 'assets/icon.png'
	}]), (0, _webpackBaseConfig.exists)(src('assets')) && {
		from: 'assets',
		to: 'assets'
	}].filter(Boolean)), new _pushManifest2.default()]), htmlPlugin(env, src('.')), isProd ? production(env) : development(env), (0, _webpack.customConfig)({
		resolveLoader: {
			alias: {
				'async': (0, _path.resolve)(__dirname, './async-component-loader')
			}
		}
	})].filter(Boolean));
};

const development = config => {
	let port = process.env.PORT || config.port || 8080,
	    host = process.env.HOST || config.host || '0.0.0.0',
	    origin = `${config.https === true ? 'https' : 'http'}://${host}:${port}/`;

	return (0, _webpack.group)([(0, _webpack.addPlugins)([new _webpack.webpack.NamedModulesPlugin()]), (0, _devServer2.default)({
		port,
		host,
		inline: true,
		hot: true,
		https: config.https,
		compress: true,
		publicPath: '/',
		contentBase: (0, _path.resolve)(config.cwd, config.src || './src'),

		disableHostCheck: true,
		historyApiFallback: true,
		quiet: true,
		clientLogLevel: 'none',
		overlay: false,
		stats: 'minimal',
		watchOptions: {
			ignored: [(0, _path.resolve)(config.cwd, 'build'), (0, _path.resolve)(config.cwd, 'node_modules')]
		}
	}, [`webpack-dev-server/client?${origin}`, `webpack/hot/dev-server?${origin}`])]);
};

const production = config => (0, _webpack.addPlugins)([new _webpack.webpack.optimize.UglifyJsPlugin({
	output: {
		comments: false
	},
	mangle: true,
	sourceMap: true,
	compress: {
		properties: true,
		keep_fargs: false,
		pure_getters: true,
		collapse_vars: true,
		warnings: false,
		screw_ie8: true,
		sequences: true,
		dead_code: true,
		drop_debugger: true,
		comparisons: true,
		conditionals: true,
		evaluate: true,
		booleans: true,
		loops: true,
		unused: true,
		hoist_funs: true,
		if_return: true,
		join_vars: true,
		cascade: true,
		drop_console: false,
		pure_funcs: ['classCallCheck', '_classCallCheck', '_possibleConstructorReturn', 'Object.freeze', 'invariant', 'warning']
	}
}), new _swPrecacheWebpackPlugin2.default({
	filename: 'sw.js',
	navigateFallback: 'index.html',
	navigateFallbackWhitelist: [/^(?!\/__).*/],
	minify: true,
	stripPrefix: config.cwd,
	staticFileGlobsIgnorePatterns: [/polyfills(\..*)?\.js$/, /\.map$/, /push-manifest\.json$/]
})]);

const htmlPlugin = (config, src) => {
	const htmlWebpackConfig = ({ url, title }) => ({
		filename: (0, _path.resolve)(config.dest, url.substring(1), 'index.html'),
		template: `!!ejs-loader!${config.template || (0, _path.resolve)(__dirname, '../../resources/template.html')}`,
		minify: config.production && {
			collapseWhitespace: true,
			removeScriptTypeAttributes: true,
			removeRedundantAttributes: true,
			removeStyleLinkTypeAttributes: true,
			removeComments: true
		},
		favicon: (0, _webpackBaseConfig.exists)((0, _path.resolve)(config.src, 'assets/favicon.ico')) ? 'assets/favicon.ico' : (0, _path.resolve)(__dirname, '../../resources/favicon.ico'),
		manifest: config.manifest,
		inject: true,
		compile: true,
		preload: config.preload === true,
		title: title || config.title || config.manifest.name || config.manifest.short_name || (config.pkg.name || '').replace(/^@[a-z]\//, '') || 'Preact App',
		excludeAssets: [/(bundle|polyfills)(\..*)?\.js$/],
		config,
		ssr(params) {
			return config.prerender ? (0, _prerender2.default)({ dest: config.dest, src }, _extends({}, params, { url })) : '';
		}
	});
	const pages = (0, _webpackBaseConfig.readJson)((0, _path.resolve)(config.cwd, config.prerenderUrls || '')) || [{ url: "/" }];
	return (0, _webpack.addPlugins)(pages.map(page => new _htmlWebpackPlugin2.default(htmlWebpackConfig(page))).concat([new _htmlWebpackExcludeAssetsPlugin2.default(), new _scriptExtHtmlWebpackPlugin2.default({
		defaultAttribute: 'defer'
	})]));
};
